OpInfo: Remove promotes_integers_to_float and infer it instead

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50438 Support JIT shape analysis for div with rounding_mode argument
* #50437 Sparse support for division rounding_mode argument
* #50281 Warn about floor_divide performing incorrect rounding
* #50280 Add division overload with rounding_mode selection
* **#50279 OpInfo: Remove promotes_integers_to_float and infer it instead**

This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

Differential Revision: [D25950011](https://our.internmc.facebook.com/intern/diff/D25950011)

peterbell10 committed:
OpInfo: Remove promotes_integers_to_float and infer it instead

This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -58,6 +58,16 @@ def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=N
+            (f'output_process_fn_grad={self.output_process_fn_grad}'
+             if self.output_process_fn_grad is not None else None)]
+
+        return f'SampleInput({", ".join(a for a in arguments if a is not None)})'
```
This is slightly out of place, but it's useful for debugging `test_ops.py` test failures because you can see which sample is being run in the pytest traceback.

e.g. `SampleInput(input[1], args=(2,), kwargs={'rounding_mode': 'floor'})`

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

mruberry commented:
Implementing dunder repr for SampleInputs and more metadata for whether the function implements safe casting sounds good. Do we really want to get rid of the metadata for whether the operator promotes integers to float, though? That seems useful.

I understand that with the changes to torch.divide(...) we have a function that conditionally promotes integers to floating point. Is a better way to model this function that torch.divide(..., mode=X) is a different function from torch.divide(..., mode=Y), though?

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

peterbell10 commented:
Note that either way, `promote_integers_to_float` is not general enough to make these type of tests work for all operators. For example, `torch.fft.fft` promotes integers to complex not float and `torch.count_nonzero` returns long from bool inputs. The approach I just pushed of calling `c10::canCast` on the other hand is completely general.

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

mruberry commented:
> Note that either way, `promote_integers_to_float` is not general enough to make these type of tests work for all operators. For example, `torch.fft.fft` promotes integers to complex not float and `torch.count_nonzero` returns long from bool inputs. The approach I just pushed of calling `c10::canCast` on the other hand is completely general.

Cool; I'll take a look. Another option would be to provide sugars for common cases (like int to float promotion), but also allowing OpInfos to override with an arbitrary mapping of inputs -> expected result dtype.

mruberry commented on test/test_ops.py:
```diff
@@ -215,8 +215,7 @@ def test_variant_consistency_eager(self, device, dtype, op):
                 #   on tensors with integer dtypes.
-                if (variant is inplace and op.promotes_integers_to_float and
-                        dtype in (torch.bool, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64)):
+                if (variant is inplace and not torch._C._can_cast(expected_forward.dtype, dtype)):
```
We can just call the torch.can_cast here, right?

https://pytorch.org/docs/master/generated/torch.can_cast.html?highlight=can_cast#torch.can_cast

peterbell10 replied:
Ah, I hadn't realised it was already exported to python.

mruberry commented on torch/_C/__init__.pyi.in:
```diff
@@ -502,6 +502,7 @@ def _vmapmode_increment_nesting() -> _int: ...  # THPModule_vmapmode_increment_n
 def _vmapmode_decrement_nesting() -> _int: ...  # THPModule_vmapmode_decrement_nesting
 def _log_api_usage_once(str) -> None: ...  # LogAPIUsageOnceFromPython
 def _demangle(str) -> str: ...  # c10::demangle
+def _can_cast(from_dtype: torch.dtype, to_dtype: torch.dtype) -> bool: ... # c10::canCast
```
See above about using the existing function exposed to Python.

mruberry commented on torch/csrc/Module.cpp:
```diff
@@ -859,6 +859,15 @@ Call this whenever a new thread is created in order to propagate values from
     }
   );
 
+  py_module.def("_can_cast", [](py::object from, py::object to) {
```
Same as above, this probably won't be necessary?

mruberry commented:
From our conversation and looking at the latest I'm convinced. It is a bit of a pain that the op has to be run to determine the output dtype, but c'est la vie. I think, however, we can use torch.can_cast and that should fix the failing builds.

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

peterbell10 commented:
@mruberry PTAL, `torch.can_cast` did indeed fix CI.

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

[ghstack-poisoned]

mruberry reviewed (approved):
Cool!

mruberry commented:
Hey @peterbell10, unfortunately the internal tools can't figure out to merge. I know these rebases are out of control, but would you mind rebasing this so I can reimport it?

peterbell10 committed:
Update on "OpInfo: Remove promotes_integers_to_float and infer it instead"


This allows different sample inputs to have different behavior for the same
operator. For example, `div(..., rounding_mode='true')` will promote but other
rounding modes don't. The current boolean flag is too restrictive to allow this.

Differential Revision: [D25950011](https://our.internmc.facebook.com/intern/diff/D25950011)

[ghstack-poisoned]

peterbell10 commented:
@mruberry rebased.
